File: paranthesis.java
/* Time complexity: O(n)
space complexity: O(n)
*/

class Solution{
    public boolean isValid(String s){
        Stack<Character> st = new Stack<>();
        for(int i=0; i<s.length(); i++){
            char c = s.charAt(i);
            if(c == '(') st.push(')');
            else if(c == '[') st.push(']');
            else if(c == '{') st.push('}');
            else if(st.isEmpty() || st.pop() != c) return false;


        }
    }
    return st.isEmpty();
}

File: Paranthesis.py
class Solution:
    def isValid(self, s: str) -> bool:
        pairs = { "(":")", "{":"}", "[":"]"}
        stack = []
        
        for i in s:
            if i in pairs:
                stack.append(i)
            elif stack:
                if pairs[stack[-1]] != i:
                    return False
                else:
                    stack.pop()
            else:
                return False
        return not stack
Time: O(n)
SPace: O(n)
